### DISCUSSION TITLE
Run Artisan Command on Lambda Function

### DISCUSSION THREAD

**Discussion by h-kenji1** [brefphp/bref#1892] [Support] [answered]
Hey, guys!

I've been trying to run an artisan command through the AWS CLI, but it doesn't work at all.
Here's my CDK Code:

    new ConsoleFunction(this, 'ArtisanLambdaFunction', {
      functionName: 'dev-artisan',
      code: packagePhpCode(`${backendProjectDir}`, {
        exclude: ['docs'],
      }),
      handler: 'artisan',
      phpVersion: '8.2'
    });

The php-82 and console layer seem to be set correctly.
After running the command:
aws lambda invoke --function-name dev-artisan --cli-binary-format raw-in-base64-out  --payload '"artisan list"' response.json

I get the error:

An error occurred (InvalidRequestContentException) when calling the Invoke operation: Could not parse request body into json: Could not parse payload into json: Unrecognized token 'route': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (byte[])"route:list"; line: 1, column: 7]

But if I execute the following command:
aws lambda invoke --function-name dev-artisan --cli-binary-format raw-in-base64-out  --payload '"{}"' response.json

This is my answer:
{
    "StatusCode": 200,
    "ExecutedVersion": "$LATEST"
}

It seems that artisan/Lambda is configured correctly, but maybe the artisan command syntax needs to be adjusted..
 Does anyone have an idea on how to solve this?

**Comment by mnapoli** [comment]
Hi,

> Could not parse payload into json: Unrecognized token 'route': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')

That tells me that the payload sent to Lambda is not recognized as a JSON string. I.e. `--payload '"artisan list"'` somehow the double quotes are not enough.

Maybe some quotes need to be escaped with backslashes or something like this. I'm surprised that this works differently than `serverless invoke` but it seems to be the case for a reason I ignore.

**Comment by h-kenji1** [comment]
Hey! Thanks for the answer!

> That tells me that the payload sent to Lambda is not recognized as a JSON string. I.e. --payload '"artisan list"' somehow the double quotes are not enough.

Yeah, that's what I've been thinking about and I actually tried everything that came up to mind and still get those errors :(

Do you have any suggestions on how should I try to execute the artisan commands?

**Reply by mnapoli** [reply]
This is deprecated but you could try sending a JSON object with a `cli` key:

```json
{"cli":"list --verbose"}
```

See https://github.com/brefphp/bref/blob/master/src/ConsoleRuntime/Main.php#L38

Also note that you shouldn't include `artisan` in the command since it's already the handler.

**Reply by h-kenji1** [reply, marked as answer]
Hey, man! Thanks one more time!
I finally could solve it!

Actually, the cli key is necessary indeed but it also needs to be escaped with backslashes like you mentioned previously.

This is it how worked for me:
`aws lambda invoke --function-name functionName --cli-binary-format raw-in-base64-out --payload '{ \"cli\":\"route:list\"}' out`

The bref documentation says:(https://bref.sh/docs/runtimes/console)
`Note: The --payload needs to contain a JSON string, that is why it is quoted twice: '"..."'. This is intentional.`

So maybe the command itself doesn't need  to be quoted twice anymore I suppose? I'm not sure..
Anyway, thanks for the help!

**Reply by mnapoli** [reply]
ohh then maybe `--payload '\"artisan list\"'` should work then.

**Reply by h-kenji1** [reply]
Unfortunately it didn't work :(.

I got the error: Unknown options: response.json

So I tried to add "" to the command itself, like `"'\"artisan list\"'"` and ended up getting the same errors I had before..
